### PR TITLE
Make check_samplesheet.py compatible with nf-core/fetchngs

### LINKED ISSUE #89: samplesheet_check.py is not compatible with fetchngs generated output
### Description of the bug

samplesheet_check.py currently forces a specific set of headers and the order of the header.

This is not compatible with fetchngs pipeline-specific output which includes a bunch of other metadata and also in it's 'own' column ordering, with the request for the user to 'double check' input and rename columns where necessary with contents of other columns. 

I think it would be good that we can just directly accept the output from fetchngs (with the different order/extra columns), for cases where all the downloaded data is assumed correct (which hopefully is most cases).

Furthermore, loosening the order and allowing extra columns would support additoinal use cases e.g. @sofstam & co who would like to record case/control-like information.

More specifically:

https://github.com/nf-core/taxprofiler/blob/6c6f5e0913add17129fcd4c6155212a9dc0cd5e1/bin/check_samplesheet.py#L81-L96

Is the error that is reported when using the fetchngs output from https://github.com/nf-core/fetchngs/pull/97 

We should just check that the REQUIRED columns are present only

### Command used and terminal output

_No response_

### Relevant files

_No response_

### System information

_No response_

### PR DESCRIPTION
**Pull request by jfy133**
nf-core/fetchNGS produces samplesheets with many more extra metadata columns than we require, and also out of order from what we currently 'expect' during our samplesheet check.

We should make sure we check only for the mandatory columns (to allow metadata), and not necessarily require a fixed 'order' of columns. This makes it more compatible with single 'unified' samplesheets for whole projects that may want more metadata, as well as fetchngs.

This PR sees me attempting to 'do python' to meet these two criteria.

Close #89 

## PR checklist

- [x] This comment contains a description of changes (with reason).
- [ ] If you've fixed a bug or added code that should be tested, add tests!
  - [ ] If you've added a new tool - have you followed the pipeline conventions in the [contribution docs](https://github.com/nf-core/taxprofiler/tree/master/.github/CONTRIBUTING.md)
  - [ ] If necessary, also make a PR on the nf-core/taxprofiler _branch_ on the [nf-core/test-datasets](https://github.com/nf-core/test-datasets) repository.
- [x] Make sure your code lints (`nf-core lint`).
- [ ] Ensure the test suite passes (`nextflow run . -profile test,docker --outdir <OUTDIR>`).
- [ ] Usage Documentation in `docs/usage.md` is updated.
- [ ] Output Documentation in `docs/output.md` is updated.
- [ ] `CHANGELOG.md` is updated.
- [ ] `README.md` is updated (including new tool citations and authors/contributors).

### CODE REVIEW

**Comment by jfy133** [comment]
The first commit replaced the original 'column order' check with a check making sure that any mandatory columns are not missing.

However `## Check sample name entries` (lines 123-134) seems to again require the 'fixed order' of the column names, so this is not yet fully resolved. I wonder if there is a way to convert the header/each row, as a named list, and pick these the required columns for checking (similar to Groovy meta maps we are using in nf-core modules).

Any fast ideas @maxibor or @Midnighter ? Can this  be done in native python or would it require libraries?

**Comment by jfy133** [comment]
Probably terrible python and violations of loads of best practises, but it should 'work' [TM]

File: bin/check_samplesheet.py [hunk 1]
```diff
@@ -87,17 +88,28 @@ def check_samplesheet(file_in, file_out):
...
-                    ",".join(header), ",".join(HEADER)
+                "ERROR: Missing required column header -> {}. Note some columns can otherwise be empty. See pipeline documentation (https://nf-co.re/taxprofiler/usage).".format(
+                    ",".join(missing_columns)
                 )
             )
             sys.exit(1)
 
+        ## Find locations of mandatory columns
+        header_locs = dict()
+        for i in HEADER:
```
Comment (sofstam) [inline bin/check_samplesheet.py:104]: That's more of a question but if one would like to add an extra column, can one add the name header as well? I am thinking that `HEADER` is already predefined. 

Like:

`sample, run_accession, instrument_platform, fastq_1, fastq_2, fasta, control`

Comment (jfy133) [reply bin/check_samplesheet.py:104]: HEADER is meant to be the minimum mandatory columns we require. So yes we could extend the list in HEADER to add more mandatory columns if necessary for the pipeline. 

But all other columns are basically 'ignored' for checking here (which is what I'm sort of trying to do in this PR). The output from the tool should only include the columns we need for the pipeline to be executed.

**Comment by maxibor** [comment]
> The first commit replaced the original 'column order' check with a check making sure that any mandatory columns are not missing.
> 
> However `## Check sample name entries` (lines 123-134) seems to again require the 'fixed order' of the column names, so this is not yet fully resolved. I wonder if there is a way to convert the header/each row, as a named list, and pick these the required columns for checking (similar to Groovy meta maps we are using in nf-core modules).
> 
> Any fast ideas @maxibor or @Midnighter ? Can this be done in native python or would it require libraries?

What you're looking for is DictReader from the csv library in base Python https://docs.python.org/3/library/csv.html#csv.DictReader

**Comment by jfy133** [comment]
> > The first commit replaced the original 'column order' check with a check making sure that any mandatory columns are not missing.
> > However `## Check sample name entries` (lines 123-134) seems to again require the 'fixed order' of the column names, so this is not yet fully resolved. I wonder if there is a way to convert the header/each row, as a named list, and pick these the required columns for checking (similar to Groovy meta maps we are using in nf-core modules).
> > Any fast ideas @maxibor or @Midnighter ? Can this be done in native python or would it require libraries?
> 
> What you're looking for is DictReader from the csv library in base Python https://docs.python.org/3/library/csv.html#csv.DictReader

Aha! Awesome! Would that be more robust than what I added in the end?

**Comment by jfy133** [comment]
I'm a bit loathe to have to re-write the entire script now. So if I can get a second OK for my 'solution' for this, it would be great.

Note that fetchngs with taxprofiler support is now available in the latest fetchngs release, so it would be good to get this in ASAP

File: bin/check_samplesheet.py [hunk 2]
```diff
@@ -87,17 +88,28 @@ def check_samplesheet(file_in, file_out):
...
-                "ERROR: Please check samplesheet header -> {} != {}".format(
-                    ",".join(header), ",".join(HEADER)
+                "ERROR: Missing required column header -> {}. Note some columns can otherwise be empty. See pipeline documentation (https://nf-co.re/taxprofiler/usage).".format(
+                    ",".join(missing_columns)
                 )
             )
             sys.exit(1)
 
+        ## Find locations of mandatory columns
+        header_locs = dict()
```
Comment (Midnighter) [inline bin/check_samplesheet.py:103]:
```suggestion
        header_locs = {}
```

**Review by Midnighter — APPROVED** [review]
It's missing some of the niceties of the updated `check_samplesheet.py` script (like allowing TSV and CSV input) but it _should_ work as intended. I agree with not putting more work into it since we can hopefully replace it soon.

**Pull request merged**